### PR TITLE
Fetch certificates for revocations

### LINKED ISSUE #1545: Python: fetch trcs & cert chains for verifying SignedCtrlPlds/SignedRevInfos/etc
Currently `scion_elem` has support for deferring the processing of a path segment until the required TRCs and cert chains have been acquired. The same approach is needed for signed ctrl payloads and signed rev infos. For signed ctrl payloads this needs to be conditional on the contents of the payload (e.g. don't try to verify payload signatures for trc/cert chain updates), and for rev info this needs to be tied to the ttl of the rev info (there's no point in trying to indefinitely verify a rev info that has expired).

### PR DESCRIPTION
**Pull request by worxli**
Fixes #1545 

This change is [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)

### CODE REVIEW

**Comment by kormat** [comment]
Review status: 0 of 6 files reviewed, 4 unresolved discussions (waiting on @worxli)

---

*[python/scion_elem/scion_elem.py, line 214 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtMF3o2FhDwhpM5-a_:-LEtMF3o2FhDwhpM5-aa:b-ueg7rj) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L214)):*
> ```Python
>         self.unv_certs = {}
>         self.unv_certs_lock = threading.Lock()
>         self.cert_reqs = {}
> ```

If you have a dict of keys to collections (e.g. IA->[list, of, requests]), then use a `defaultdict(<collection>)`. See `_core_ases` as example.

---

*[python/scion_elem/scion_elem.py, line 479 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEt84Ga9_gqjwDr54Jz:-LEt84GbBBWqAtd5EBHu:b-as4tkd) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L479)):*
> ```Python
>         for isd, ver in missing_trcs:
>             with self.req_trcs_lock:
>                 req_time, meta = self.requested_trcs.get((isd, ver), (None, None))
> ```

This block of logic should be in `_request_trc()` instead.

---

*[python/scion_elem/scion_elem.py, line 522 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtOxX12zU_VTsoMxvR:-LEtOxX12zU_VTsoMxvS:b-mdva9b) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L522)):*
> ```Python
>         for isd_as, ver in missing_certs:
>             with self.req_certs_lock:
>                 req_time, meta = self.requested_certs.get((isd_as, ver), (None, None))
> ```

This block of logic should be in `_request_cert`.

---

*[python/scion_elem/scion_elem.py, line 1341 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtP7F0BaITdKouHImU:-LEtP7F0BaITdKouHImV:b-6mefom) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L1341)):*
> ```Python
>             self.verify_revocation(src, srev_info, callback)
> 
>     def wait_for_cert(self, src, srev_info, callback):
> ```

The name is generic, but apparently this is specific to revocations?

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 0 of 8 files reviewed, 4 unresolved discussions (waiting on @kormat)

---

*[python/scion_elem/scion_elem.py, line 214 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtMF3o2FhDwhpM5-a_:-LExYny9BEaRqAXpV2_x:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L214)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

If you have a dict of keys to collections (e.g. IA->[list, of, requests]), then use a `defaultdict(<collection>)`. See `_core_ases` as example.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 479 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEt84Ga9_gqjwDr54Jz:-LExYp8iBS5w293XsnLE:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L479)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This block of logic should be in `_request_trc()` instead.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 522 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtOxX12zU_VTsoMxvR:-LExYrgU6M24RpnBnP38:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L522)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This block of logic should be in `_request_cert`.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 1341 at r1](https://reviewable.io/reviews/scionproto/scion/1600#-LEtP7F0BaITdKouHImU:-LExYttnELkH4p79r713:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cb44076053bc103eb1d4768f5035da2b3066d0aa/python/scion_elem/scion_elem.py#L1341)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

The name is generic, but apparently this is specific to revocations?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Reviewed 1 of 6 files at r1, 7 of 7 files at r2.
Review status: all files reviewed, 8 unresolved discussions (waiting on @worxli)

---

*[python/lib/packet/cert_mgmt.py, line 33 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEycJbO5gJWR-jT8MQs:-LEycJbO5gJWR-jT8MQt:be9d0ir) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/lib/packet/cert_mgmt.py#L33)):*
> ```Python
> 
> 
> class CertRequest(object):
> ```

`CertRequestState` would be a better name.

---

*[python/scion_elem/scion_elem.py, line 402 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyV2qu1dorqprjVuKV:-LEyV2qu1dorqprjVuKW:b-dk7zjm) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L402)):*
> ```Python
>                 # Try to find a valid meta, otherwise the CS is asked
>                 meta = None
>                 for cert_req in cert_reqs:
> ```

Assuming this list is appended to, i would search from the end, to find the newest meta.

---

*[python/scion_elem/scion_elem.py, line 546 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyWpV_235f5phE7LDK:-LEyWpV_235f5phE7LDL:b8vnh5a) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L546)):*
> ```Python
>             logging.error("Couldn't find a CS to request %sv%s CERTCHAIN" % (isd_as, ver))
>             return
>         cert_req = CertChainRequest.from_values(isd_as, ver, cache_only=True)
> ```

I'm not sure that `cache_only` is desirable. If the CS doesn't have the required trc/cert, we want it to fetch it.

---

*[python/scion_elem/scion_elem.py, line 686 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXe383N0pS8b3VZom:-LEyXe383N0pS8b3VZon:b-88ttgv) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L686)):*
> ```Python
>         if not trc:
>             # TRC still being requested
>             with self.unv_certs_lock:
> ```

I think this needs to be an RLock.

---

*[python/scion_elem/scion_elem.py, line 699 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXi-13LByBr3xMJd8:-LEyXi-13LByBr3xMJd9:beenn09) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L699)):*
> ```Python
>             self.unv_certs.pop((isd_as, ver), None)
>         # Remove received cert chain from map
>         self._check_segs_with_rec_cert(isd_as, ver)  # TODO handle with callbacks / events
> ```

`TODO(worxli)`

---

*[python/scion_elem/scion_elem.py, line 704 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXngU0QNHv5lQyiji:-LEyXngU0QNHv5lQyijj:beenn09) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L704)):*
> ```Python
>         # Got a TRC, thus check all unverified certificates
>         with self.unv_certs_lock:
>             for (isd_as, ver), cert in self.unv_certs.items():  # TODO check if save
> ```

`TODO(worxli)`

---

*[python/scion_elem/scion_elem.py, line 1381 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEybuLRCk019ZYGprvw:-LEybuLRCk019ZYGprvx:b-wl26l0) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L1381)):*
> ```Python
>             srev_info.verify(cert.as_cert.subject_sig_key_raw)
>         except SCIONBaseError as e:
>             logging.error("Revocation check failed for %s:\n%s from %s",
> ```

s/check/verification/

---

*[python/scion_elem/scion_elem.py, line 1416 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEya0dE4RNZJYvT7Z4M:-LEya0dE4RNZJYvT7Z4N:b-km719c) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L1416)):*
> ```Python
>         cert_req = CertRequest(src, meta)
>         with self.cert_reqs_lock:
>             cert_reqs = self.cert_reqs.pop((src.ia, src.chain_ver), list())
> ```

`self.cert_reqs[(src.ia, src.chain_ver)].append(cert_reqs)`
This is the advantage of using a `defaultdict(list)`

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 6 of 8 files reviewed, 8 unresolved discussions (waiting on @kormat)

---

*[python/lib/packet/cert_mgmt.py, line 33 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEycJbO5gJWR-jT8MQs:-LF2Ldh75UfP_7CyZVoE:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/lib/packet/cert_mgmt.py#L33)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`CertRequestState` would be a better name.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 402 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyV2qu1dorqprjVuKV:-LF2M8CQ8gXpPCxxYtF5:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L402)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Assuming this list is appended to, i would search from the end, to find the newest meta.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 546 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyWpV_235f5phE7LDK:-LF2MJBv2NbdfyXx9nct:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L546)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I'm not sure that `cache_only` is desirable. If the CS doesn't have the required trc/cert, we want it to fetch it.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 686 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXe383N0pS8b3VZom:-LF2M_VS-GmG-nmKK1fE:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L686)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I think this needs to be an RLock.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 699 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXi-13LByBr3xMJd8:-LF2NtFK9M16iiFkHVlE:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L699)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`TODO(worxli)`
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 704 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEyXngU0QNHv5lQyiji:-LF31pBe5xQ8qS1GDY-M:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L704)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`TODO(worxli)`
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 1381 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEybuLRCk019ZYGprvw:-LF2Mghp1GVmfUh_G2A0:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L1381)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

s/check/verification/
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 1416 at r2](https://reviewable.io/reviews/scionproto/scion/1600#-LEya0dE4RNZJYvT7Z4M:-LF2MrvnD5qZld3KMeh4:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/08f1bcad4944a5603e0631bff2e994369bf338c6/python/scion_elem/scion_elem.py#L1416)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`self.cert_reqs[(src.ia, src.chain_ver)].append(cert_reqs)`
This is the advantage of using a `defaultdict(list)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Reviewed 1 of 2 files at r3.
Review status: 7 of 8 files reviewed, 3 unresolved discussions (waiting on @kormat and @worxli)

---

*[python/scion_elem/scion_elem.py, line 414 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSVFb31iws5l0KfWWc:-LFSVFb31iws5l0KfWWd:b-4lhe3e) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L414)):*
> ```Python
>                         break
>                 # Finally register the necessary requests
>                 trc = self.trust_store.get_trc(isd_as[0], src.trc_ver)
> ```

```
if not self.trust_store.get_trc(isd_as[0], src.trc_ver):
   # TRC must also be fetched
   ...
```
as `trc` is never used again.

---

*[python/scion_elem/scion_elem.py, line 502 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSXYD-5_X9B7MP8-oN:-LFSXYD0-G-2KkIgWDD7:bfyb9tf) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L502)):*
> ```Python
>         with self.req_trcs_lock:
>             req_time, req_meta = self.requested_trcs.get((isd, ver), (None, None))
>             if req_meta and meta:
> ```

Proposed change:
```
            req_time, req_meta = self.requested_trcs.get((isd, ver), (None, None))
            if req_time:
                if meta:
                    # There is already an outstanding request for the missing TRC
                    # Update the stored meta with the latest known server that has the TRC.
                    self.requested_trcs[(isd, ver)] = (req_time, meta)
                logging.debug("Request for %sv%s TRC already registered" % (isd, ver))
                return
```

---

*[python/scion_elem/scion_elem.py, line 507 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSVmz_ArGO664PTIjT:-LFSVmz_ArGO664PTIjU:b-c60mvi) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L507)):*
> ```Python
>                 self.requested_trcs[(isd, ver)] = (req_time, meta)
>             if req_meta:
>                 logging.debug("Request for %sv%s TRC already registered" % (isd, ver))
> ```

Include the meta (if any), for debugging.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Review status: 7 of 8 files reviewed, 6 unresolved discussions (waiting on @kormat and @worxli)

---

*[python/scion_elem/scion_elem.py, line 337 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX-weMElGF9CseasSn:-LFX-weN74Z3VpbDRC0E:b9ll38t) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L337)):*
> ```Python
>             sleep_interval(start, check_cyle, "Elem._check_trc_cert_reqs cycle")
> 
>     def _process_cert_reqs(self):
> ```

It _should_ be possible to just add `self._check_cert_req_states()` into `_check_trc_cert_reqs`. I know you had issues with doing that before, but you've found/fixed some bugs since.

---

*[python/scion_elem/scion_elem.py, line 361 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX0CVF3rYjsp2cMLdU:-LFX0CVF3rYjsp2cMLdV:b-r5kugj) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L361)):*
> ```Python
>     def _send_trc_req(self, isd, ver, trc_req, req_id, meta):
>         with self.req_trcs_lock:
>             self.requested_trcs[(isd, ver)] = (time.time(), meta)
> ```

This is rather subtle - because of how this method is called it never overrides an existing meta, but you can't tell that by looking here. Maybe add a comment explaining this? Same applies to `_send_cert_req`.

---

*[python/scion_elem/scion_elem.py, line 666 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX-nhuA612n6JtwvJ2:-LFX-nhuA612n6JtwvJ3:ba9f2zr) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L666)):*
> ```Python
>         logging.info("Cert chain reply received for %sv%s from %s [id: %s]",
>                      isd_as, ver, meta, cpld.req_id_str())
>         self._verify_cert(rep.chain)
> ```

If this fails, processing should stop.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Review status: 7 of 8 files reviewed, 6 unresolved discussions (waiting on @kormat and @worxli)

---

*[python/scion_elem/scion_elem.py, line 361 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX0CVF3rYjsp2cMLdU:-LFX7tcG6lTD4vwGJpTz:bxw7904) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L361)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is rather subtle - because of how this method is called it never overrides an existing meta, but you can't tell that by looking here. Maybe add a comment explaining this? Same applies to `_send_cert_req`.
</blockquote></details>

Actually, this shouldn't insert the new meta regardless. `_get_cs()` can return different CS instances over time. If a CS instance is removed, having the old meta here will prevent a new CS from being retrieved in certain circumstances.

TL;DR:
- do not store a new `meta` here, only update the timstamp.
- have `meta = meta or self._get_cs()` in this function, not in `_check_trc_reqs`

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Reviewed 1 of 2 files at r3.
Review status: all files reviewed, 7 unresolved discussions (waiting on @worxli)

---

*[python/path_server/base.py, line 250 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFXNgwJA_rwrsCALBdb:-LFXNgwJA_rwrsCALBdc:btethha) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/path_server/base.py#L250)):*
> ```Python
>         logging.debug("Received revocation from %s: %s", meta, rev_info.short_desc())
>         if meta.ia[0] != self.addr.isd_as[0]:
>             logging.info("Dropping revocation received from a different ISD. Src: %s SRevInfo: %s" %
> ```

Hmm. What's the logic for this?

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @kormat and @worxli)

---

*[python/path_server/base.py, line 250 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFXNgwJA_rwrsCALBdb:-LFXdiEP2j2HzmuXmTti:b-cyeina) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/path_server/base.py#L250)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Hmm. What's the logic for this?
</blockquote></details>

Why it is dropped? You may ask @shitz who introduced that in #947.

---

*[python/scion_elem/scion_elem.py, line 337 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX-weMElGF9CseasSn:-LFXeCot2M3TZBnA_IHi:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L337)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It _should_ be possible to just add `self._check_cert_req_states()` into `_check_trc_cert_reqs`. I know you had issues with doing that before, but you've found/fixed some bugs since.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 414 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSVFb31iws5l0KfWWc:-LFXf-7y0QAcfI7_3RPd:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L414)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

```
if not self.trust_store.get_trc(isd_as[0], src.trc_ver):
   # TRC must also be fetched
   ...
```
as `trc` is never used again.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 666 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX-nhuA612n6JtwvJ2:-LFXfu5K3hKyw3VH2os3:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L666)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

If this fails, processing should stop.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @kormat and @worxli)

---

*[python/scion_elem/scion_elem.py, line 502 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSXYD-5_X9B7MP8-oN:-LFXhdUD2SiHiJrh0QxH:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L502)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Proposed change:
```
            req_time, req_meta = self.requested_trcs.get((isd, ver), (None, None))
            if req_time:
                if meta:
                    # There is already an outstanding request for the missing TRC
                    # Update the stored meta with the latest known server that has the TRC.
                    self.requested_trcs[(isd, ver)] = (req_time, meta)
                logging.debug("Request for %sv%s TRC already registered" % (isd, ver))
                return
```
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 507 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFSVmz_ArGO664PTIjT:-LFXhlSMEagje4xMx8AY:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L507)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Include the meta (if any), for debugging.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @kormat)

---

*[python/scion_elem/scion_elem.py, line 361 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX0CVF3rYjsp2cMLdU:-LFXoyxKBd0C5T6qwyl3:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L361)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Actually, this shouldn't insert the new meta regardless. `_get_cs()` can return different CS instances over time. If a CS instance is removed, having the old meta here will prevent a new CS from being retrieved in certain circumstances.

TL;DR:
- do not store a new `meta` here, only update the timstamp.
- have `meta = meta or self._get_cs()` in this function, not in `_check_trc_reqs`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by worxli** [comment]
Review status: 4 of 8 files reviewed, 7 unresolved discussions (waiting on @kormat)

---

*[python/scion_elem/scion_elem.py, line 361 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX0CVF3rYjsp2cMLdU:-LFXp8DLEnM3lwVaiCKv:btb8q52) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L361)):*
<details><summary><i>Previously, worxli (Lukas Bischofberger) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Meta needs to be stored to begin with. When re-requesting the same meta is put again. And with having `meta = meta or self._get_cs()` after the update, we make sure that a CS meta is never put into the request list.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Reviewed 4 of 4 files at r4.
Review status: all files reviewed, 2 unresolved discussions (waiting on @kormat)

---

*[python/scion_elem/scion_elem.py, line 361 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFX0CVF3rYjsp2cMLdU:-LFY2F-IAQmZnYHWYRkF:bjdf19e) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/scion_elem/scion_elem.py#L361)):*
<details><summary><i>Previously, worxli (Lukas Bischofberger) wrote…</i></summary><blockquote>

Meta needs to be stored to begin with. When re-requesting the same meta is put again. And with having `meta = meta or self._get_cs()` after the update, we make sure that a CS meta is never put into the request list.
</blockquote></details>

I don't follow.
If `_request_trc` is called with a non-`None` meta, then that should be used. This is already handled by `_request_trc`.
If it's called with a `None` meta, and there's no existing entry in `requested_trcs`, then we should store `None` in there, _not_ the output of `_get_cs()`.

Let's talk tomorrow.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
:lgtm:

---

Review status: all files reviewed, 1 unresolved discussion (waiting on @kormat)

---



*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600#-:-LFamRbdEU5yHdXEGGPP:bnfp4nl)*

**Comment by kormat** [comment]
Review status: :shipit: complete! all files reviewed, all discussions resolved

---

*[python/path_server/base.py, line 250 at r3](https://reviewable.io/reviews/scionproto/scion/1600#-LFXNgwJA_rwrsCALBdb:-LFanUueDS92LxrMY9OY:biqc6ik) ([raw file](https://github.com/scionproto/scion/blob/cf2dcc92acfbffd8f13e207442088437d6205946/python/path_server/base.py#L250)):*
<details><summary><i>Previously, worxli (Lukas Bischofberger) wrote…</i></summary><blockquote>

Why it is dropped? You may ask @shitz who introduced that in #947.
</blockquote></details>

Ok, i can't figure out why it was added, but it's definitely wrong, so please drop this logic.

---


*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*

**Comment by kormat** [comment]
Reviewed 1 of 1 files at r5.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/scionproto/scion/1600)*
